realm/realm-library/src/androidTest/java/io/realm/internal/android/JsonUtilsTest.java

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright 2015 FasterXML
+ * Copyright 2015 Realm Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.realm.internal.android;
+
+import android.test.AndroidTestCase;
+
+import java.text.ParseException;
+import java.text.ParsePosition;
+import java.util.Calendar;
+import java.util.Date;
+import java.util.GregorianCalendar;
+import java.util.TimeZone;
+
+public class JsonUtilsTest extends AndroidTestCase {
+
+    public void testParseNullAndEmptyDateIsNull() {
+        Date output = JsonUtils.stringToDate(null);
+        assertNull("Null input should output a null date object", output);
+
+        output = JsonUtils.stringToDate("");
+        assertNull("Empty string input should output a null date object", output);
+    }
+
+    public void testParseMillisToDate() {
+        Date originalDate = Calendar.getInstance().getTime();
+        long dateTimeInMillis = originalDate.getTime();
+        Date output = JsonUtils.stringToDate(String.valueOf(dateTimeInMillis));
+
+        assertTrue("Dates should match", output.equals(originalDate));
+    }
+
+    public void testParseJsonDateToDate() {
+        String jsonDate = "/Date(1198908717056)/"; // 2007-12-27T23:11:57.056
+        Date output = JsonUtils.stringToDate(jsonDate);
+
+        assertEquals(output.getTime(), 1198908717056L);
+    }
+
+    public void testParseJsonDateShouldThrow() {
+        String invalidLongDate = "123abc";
+
+        try {
+            JsonUtils.stringToDate(invalidLongDate);
+            fail("Should fail with NumberFormatException");
+        } catch (NumberFormatException ignored) {
+            assertNotNull(ignored);
+        } catch (Exception e) {
+            fail("Should have failed with NumberFormatException not caught version.");
+        }
+    }
+
+    public void testParseISO8601Dates() throws ParseException {
+        Calendar cal = new GregorianCalendar(2007, 8 - 1, 13, 19, 51, 23);
+        cal.setTimeZone(TimeZone.getTimeZone("GMT"));
+        cal.set(Calendar.MILLISECOND, 789);
+        Date date = cal.getTime();
+        cal.set(Calendar.MILLISECOND, 0);
+        Date dateZeroMillis = cal.getTime();
+        cal.set(Calendar.SECOND, 0);
+
+        Date d = ISO8601Utils.parse("2007-08-13T195123.789Z", new ParsePosition(0));
+
+        assertEquals(date, d);
+
+        d = JsonUtils.stringToDate("2007-08-13T195123Z");
```

Comment:
Add a comment about what is being tested above each of these. Mentally parsing them is kinda hard :)